tuning/trainercontroller/callback.py

```diff
@@ -0,0 +1,353 @@
+                if rule_succeeded:
+                    for operation_action in control_action.operation_actions:
+                        logger.info(f"Taking {operation_action.action} \
+                                    action in {control_action.name}")
+                        operation_action.instance.act(action=operation_action.action, event_name=event_name, **kwargs)
+
+    def _actions_on_event(self, event_name: str, **kwargs):
+        """Invokes all functions associated with an event.
+
+        Args:
+            event_name: str. Event name.
+            kwargs: List of arguments (key, value)-pairs.
+        """
+        self._compute_metrics(event_name, **kwargs)
+        self._take_control_actions(event_name, **kwargs)
+
+    def _validate_rule(self, rule):
+        """Validates the rule to check if there are any import attempts 
+
+        Returns:
+            bool
+        """
+        return re.search(r'__', rule) is None
+    
+    def on_init_end(self, args: TrainingArguments, state: TrainerState, \
+                    control: TrainerControl, **kwargs):
+        """This event gets the training arguments which is finally used by the trainer loop. All metric and operation validation is performed here using these arguments. Following this, validated metrics and operations instances are registered for use.
+
+        Args:
+            args: TrainingArguments. Training arguments for the trainer loop.
+            state: TrainerState. Current trainer state.
+            control: TrainerControl. Trainer control object.
+            kwargs: List of arguments (key, value)-pairs.
+        """
+        # Training arguments, state and controls are folded into kwargs to be passed off to handlers
+        kwargs["args"] = args
+        kwargs["state"] = state
+        kwargs["control"] = control
+
+        # Check if there any metrics listed in the configuration
+        if CONTROLLER_METRICS_KEY not in self.trainer_controller_config:
+            logger.warn("Trainer controller config has no metrics.")
+
+        # Metric handler validation and registration is performed here.
+        for metric_name, metric_config in self.trainer_controller_config[CONTROLLER_METRICS_KEY].items():
+            # Get the metric class name from the config section.
+            metric_handler_name = next(iter(metric_config.keys()))
+            # Get the handler class using the metric class name.
+            if metric_handler_name not in self.metric_handlers:
+                raise KeyError(f"Undefined metric handler {metric_handler_name}")
+            metric_handler = self.metric_handlers[metric_handler_name]
+            # Get the metric handler class arguments specified in the config.
+            metric_args = metric_config[metric_handler_name]
+            if metric_args is None:
+                metric_args = {}
+            # Metric handler instance is created here.
+            obj = metric_handler(name=metric_name, **metric_args, **kwargs)
```

Comment:
What is the reason for passing kwargs through here? If something is written weirdly, there could be potential key collisions between that and the metric args being kwarg expanded - is there an advantage to passing it to a metric handler in this way?

Reply:
@alex-jw-brooks The reasons for the above signature in the `metric_handler` are as follows:
1. If you notice, the `metric_handler` is actually a class type that is stored following registration (every metric class type and corresponding arguments are stored and referred to using a metric name). Following code [link](https://github.com/foundation-model-stack/fms-hf-tuning/blob/7691921e97f6b827063860597ef80b5e45bb9bf6/tuning/trainercontroller/callback.py#L220) illustrates this. The  arguments could come from the trainercontroller config file. They could be any set of key-value pairs. These are passed using ` **metric_args`.
1. In addition to `**metric_args`, we also have the `**kwargs` coming from which include trainer arguments, state and controls which could potentially be used by a metric during initialization. 

Hence, we need both `**metric_args` and `**kwargs`  to remain separate as they come from different sources.

However, as you pointed out, there could be potential collision of keys between these two sources. To remedy this without compromising on the flexibility of the above approach, we will document a warning in the code as well as in the ADR, or if required add a README to inform the users to be careful in choosing the argument names.

Reply:
Makes sense, thanks for the clarification!